### ISSUE TITLE
[PERF] Why is the first `read_csv` call slower than subsequent `read_csv` calls?

### ISSUE THREAD

**Issue by YarShev** [modin-project/modin#5157]
```python
import time
import pandas
import numpy as np
import modin.pandas as pd
import modin.config as cfg
import ray

print(f"\tPandas version: {pandas.__version__}")
print(f"\tModin version: {pd.__version__}")
print(f"\tCpuCount: {cfg.CpuCount.get()}")
print(f"\tEngine: {cfg.Engine.get()}")
print(f"\tNPartitions: {cfg.NPartitions.get()}")

ray.init(runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})

pandas_df = pandas.DataFrame(
    np.random.randint(0, 100, size=(1000000, 13))
)
pandas_df.to_csv("foo.csv", index=False)

def read_csv_with_pandas():
    start_time = time.time()

    pandas_df = pandas.read_csv("foo.csv", index_col=0)

    end_time = time.time()
    pandas_duration = end_time - start_time
    print("Time to read_csv with Pandas: {} seconds".format(round(pandas_duration, 3)))
    return pandas_df


def read_csv_with_modin():
    start_time = time.time()

    modin_df = pd.read_csv("foo.csv", index_col=0)

    end_time = time.time()
    modin_duration = end_time - start_time
    print("Time to read_csv with Modin: {} seconds".format(round(modin_duration, 3))) 
    return modin_df

for i in range(5):
    pandas_df = read_csv_with_pandas()
    modin_df = read_csv_with_modin()

        Pandas version: 1.5.1
        Modin version: 0.16.0+24.g11ba4811
        CpuCount: 8
        Engine: Ray
        NPartitions: 8

Time to read_csv with Pandas: 0.708 seconds
Time to read_csv with Modin: 4.132 seconds
Time to read_csv with Pandas: 0.735 seconds
Time to read_csv with Modin: 0.37 seconds
Time to read_csv with Pandas: 0.646 seconds
Time to read_csv with Modin: 0.377 seconds
Time to read_csv with Pandas: 0.673 seconds
Time to read_csv with Modin: 0.371 seconds
Time to read_csv with Pandas: 0.672 seconds
Time to read_csv with Modin: 0.379 seconds
```

**Comment by mvashishtha** [comment]
@YarShev I have observed this behavior for a long time, and it has made benchmarking really annoying. I have typically warmed up the ray cluster before doing benchmarking by doing, e.g. an `applymap` that uses all the workers (but taking care to completely delete the warmup object so it doesn't continue to take up memory). Here's an example where the the first `applymap` is about twice as long as the subsequent ones:


```python
import modin.pandas as pd
from modin.config import BenchmarkMode
import numpy as np

BenchmarkMode.put(True)
df = pd.DataFrame(np.random.randint(0, 100, size=(2**18, 2**8)))
%time result = df.applymap(lambda x: x) # Wall time: 5.43 sec
%time result = df.applymap(lambda x: x) # Wall time: 2.91 sec
%time result = df.applymap(lambda x: x) # Wall time: 2.91 sec
```

In the past, I have seen a very high `deserialize_arguments` cost for the first call, whereas now in the ray [timeline](https://github.com/modin-project/modin/files/9870614/ray_timeline_first_slow_apply.json.zip) I see only about .6 seconds extra for `deserialize_arguments`. Maybe the difference is because back when I was looking into the last time, I was probably on an older version of ray.

@vnlitvinov has hypothesized that the initial cost comes from importing pandas.

**System information**
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS Monterey 
- **Computer model**: MacBook Pro (16-inch, 2019)
- **Memory**: 16 GB 2667 MHz DDR4
- **Processor**: 2.3 GHz 8-Core Intel Core i9
- **Modin version** (`modin.__version__`): 11ba4811e6db11740e11fd33d3cdfba8ce5bec54
- **Python version**: 3.10.6
- **Pandas version**: 1.5.1
- **Ray version**: 2.0.0

**Comment by vnlitvinov** [comment]
This is the consequence of https://github.com/modin-project/modin/pull/4603 - we've started passing non-empty `runtime_env` to `ray.init()`, which causes Ray to start workers only when there's work for them to do instead of spawning them at `ray.init()` time:

https://github.com/modin-project/modin/blob/11ba4811e6db11740e11fd33d3cdfba8ce5bec54/modin/core/execution/ray/common/utils.py#L75

What I usually do when benching is I'm adding this line before actually getting any figures:
```python
pd.DataFrame(range(cfg.CpuCount.get() * cfg.MinPartitionSize().get())).to_numpy() # init the engine and start all the workers
```

Note that just create a DataFrame isn't enough, as at creation time it's only distributed to Plasma, but, as no work is performed, no workers are spawned.

**Comment by mvashishtha** [comment]
FWIW, I have faced this issue (unfortunately not recorded in GitHub, but recorded in a private Slack) even before #4603, on March 31.

**Comment by YarShev** [comment]
Even replacing

```python
ray.init(runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})
```

to

```python
ray.init()
```

I see about 2-3x slower first `read_csv` call

```
Time to read_csv with Modin: 1.105 seconds
Time to read_csv with Pandas: 0.569 seconds
Time to read_csv with Modin: 0.358 seconds
Time to read_csv with Pandas: 0.565 seconds
Time to read_csv with Modin: 0.363 seconds
Time to read_csv with Pandas: 0.518 seconds
Time to read_csv with Modin: 0.345 seconds
Time to read_csv with Pandas: 0.535 seconds
Time to read_csv with Modin: 0.467 seconds
```

**Comment by YarShev** [comment]
That is not what the user expects to see. I wonder if @modin-project/modin-ray is aware of the problem related to not only with passing `env_vars`, but and with just `ray.init`?

At least, we should document the behavior if we can't fix it on our end.

**Comment by mvashishtha** [comment]
@YarShev I want to point out again that the bug predates the `__MODIN_AUTOIMPORT_PANDAS__` variable. I was definitely seeing the same problem long before #4603 was merged.

**Comment by YarShev** [comment]
@modin-project/modin-ray may know why this is happening.

**Comment by rkooo567** [comment]
My repro result
```
Time to read_csv with Pandas: 0.554 seconds
Time to read_csv with Modin: 3.945 seconds
Time to read_csv with Pandas: 0.716 seconds
Time to read_csv with Modin: 0.439 seconds
Time to read_csv with Pandas: 1.17 seconds
Time to read_csv with Modin: 0.506 seconds
Time to read_csv with Pandas: 1.131 seconds
Time to read_csv with Modin: 0.524 seconds
Time to read_csv with Pandas: 0.809 seconds
Time to read_csv with Modin: 0.434 seconds
```

I ran ray timeline.

<img width="1102" alt="Screen Shot 2022-10-26 at 11 20 44 PM" src="https://user-images.githubusercontent.com/18510752/198206599-fce56940-92b7-47d9-9397-a2e5dddfecfa.png">

<img width="389" alt="Screen Shot 2022-10-26 at 11 20 51 PM" src="https://user-images.githubusercontent.com/18510752/198206601-4b7f5370-a5a9-4c8c-94df-817b096bb0c9.png">

The blue one is when you submitted tasks. It seems like in the first batch of jobs, it takes a really long time until the task is scheduled and executed. I assume the overhead is the worker startup overhead. When you start a new ray job, it takes some time to start every worker. I pre-started all the workers using this. 

```
@ray.remote
def f():
    import modin # To reduce modin import overhead
    time.sleep(1)

ray.get([f.remote() for _ in range(int(ray.cluster_resources()["CPU"]))])
```

This removes the worker startup overhead, and after that most of the overhead is the deserialization of arguments (I think this is expected. Maybe the fact that the second run is faster is the result of optimization within Ray)

<img width="684" alt="Screen Shot 2022-10-26 at 11 41 54 PM" src="https://user-images.githubusercontent.com/18510752/198210408-b0ecc7bc-f00f-4518-bf33-89b39d64a205.png">

**Comment by YarShev** [comment]
@rkooo567, do you have some notes regarding this in Ray documentation? This behavior can really confuse the user.

**Comment by rkooo567** [comment]
I will update this information. I am thinking to add this information here; https://docs.ray.io/en/master/ray-observability/monitoring-debugging/troubleshoot-performance.html

**Comment by YarShev** [comment]
Can you please notify us here when your change is merged in master so we are in tune.

**Comment by YarShev** [comment]
> Yeah agreed it is confusing. I will update this information to the doc. I am thinking to add this information here; https://docs.ray.io/en/master/ray-observability/monitoring-debugging/troubleshoot-performance.html. I will add you as a reviewer

@rkooo567, any updates?

**Comment by rkooo567** [comment]
Sorry. I am kind of swamped by other TODOs... I will do my best to get this in soon

**Comment by YarShev** [comment]
@rkooo567, are there plans to fix the issue with Ray workers set up or just to add a note regarding the performance problem?

**Comment by rkooo567** [comment]
Just documentation. Worker behavior is expected (you should start worker processes before executing any task). Alternatively, we can support APIs like waiting for worker procs to start, but in that case please create a feature request!

**Comment by YarShev** [comment]
We start worker processes before executing any task but the overhead of passing `runtime_env` other than `None` in to `ray.init` is higher than it it is just `None`. Look forward to seeing a note in Ray docs.

**Comment by rkooo567** [comment]
> overhead of passing runtime_env other than None in to ray.init is higher than it it is just None

I think it is because we are creating the runtime environment if it is specified (it happens only at the first invocation). You can think if runtime env is used, you do this process: worker start -> download runtime env -> start task.

**Comment by dchigarev** [comment]
> This is the consequence of #4603 - we've started passing non-empty `runtime_env` to `ray.init()`, which causes Ray to start workers only when there's work for them to do instead of spawning them at `ray.init()` time:
> 
> https://github.com/modin-project/modin/blob/11ba4811e6db11740e11fd33d3cdfba8ce5bec54/modin/core/execution/ray/common/utils.py#L75
> 
> What I usually do when benching is I'm adding this line before actually getting any figures:
> 
> ```python
> pd.DataFrame(range(cfg.CpuCount.get() * cfg.MinPartitionSize().get())).to_numpy() # init the engine and start all the workers
> ```
> 
> Note that just create a DataFrame isn't enough, as at creation time it's only distributed to Plasma, but, as no work is performed, no workers are spawned.

So what I'm seeing is that specifying a custom runtime environment for Ray brings a noticeable overhead for starting up the workers:

| | Workers startup |
|-|-|
| When `runtime_env` is passed | 6.92s|
| When `runtime_env` isn't passed | 0.84s|

<details><summary>script to measure</summary>

```python
import modin.pandas as pd
import ray
import time
import modin.config as cfg

ray.init(
    # runtime_env={"env_vars": {"SOME_VAR": "1"}}
)

t0 = time.time()
pd.DataFrame(range(cfg.CpuCount.get() * cfg.MinPartitionSize().get())).to_numpy() # init the engine and start all the workers
print(time.time() - t0)
```

</details>

So I have a question for @vnlitvinov as you were the one who introduced this change. Is there a way to get rid of specifying this runtime env for certain cases?

As far I understand that in case of a single node execution, we can set the `__MODIN_AUTO_IMPORT_PANDAS__` env variable in the main process and it should propagate to all of the workers with no overhead (since they're running on the same machine):
```python
import ray
import os
import modin.pandas as pd
import modin.config as cfg
import time

@ray.remote
def remote_func():
    return os.environ.get("CUSTOM_ENV_VARIABLE", "default_value")

os.environ["CUSTOM_ENV_VARIABLE"] = "custom_value"
ray.init()

t0 = time.time()
pd.DataFrame(range(cfg.CpuCount.get() * cfg.MinPartitionSize().get())).to_numpy() # init the engine and start all the workers
print(time.time() - t0) # 0.86s
print(ray.get(remote_func.remote())) # actually prints 'custom_value'
```

So if this is the case, do you think the logic from your fix (#4603) will work the same if I replace passing a custom runtime env to ray with simply setting the environment variable in a single-node case?

**Comment by vnlitvinov** [comment]
Yes and no... first of all, it would affect other processes, too, and - second - there is no guarantee that all environment variables will be inherited by Ray workers, as this is implementation detail (not a contract) as far as I'm aware.

Oh, and this won't work for the case of locally running Ray to which we already connect.

With all that said, I guess you can try to do this optimization while leaving a comment pointing to this discussion, so if something breaks in the future we'd now where to start digging.
